### PR TITLE
Odometer

### PR DESCRIPTION
**Pull request by guscuizi**
Adding funtion:  transfer the odometer data  from smart car to Android APP, and display it.

### CODE REVIEW

File: android/app/src/main/java/smart/delirover/app/MailmanTraveledDistanceActivity.java [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+import androidx.appcompat.app.AppCompatActivity;
+
+import com.delirover.app.R;
+
+import org.eclipse.paho.client.mqttv3.IMqttActionListener;
+import org.eclipse.paho.client.mqttv3.IMqttDeliveryToken;
+import org.eclipse.paho.client.mqttv3.IMqttToken;
+import org.eclipse.paho.client.mqttv3.MqttCallback;
+import org.eclipse.paho.client.mqttv3.MqttMessage;
+//MQTT connection part of the code is built on https://github.com/DIT112-V21/smartcar-mqtt-controller.git
```
Comment (platisd) [inline android/app/src/main/java/smart/delirover/app/MailmanTraveledDistanceActivity.java:24]: An MQTT client [already exists](https://github.com/DIT112-V21/group-17/blob/master/android/app/src/main/java/smart/delirover/app/MqttClient.java), please reuse that.

File: arduino/smartcar/smartcar.ino [hunk 2]
```diff
@@ -85,20 +85,33 @@ void loop() {
     const auto currentTime = millis();
     #ifdef __SMCE__
         static auto previousFrame = 0UL;
-        if (currentTime - previousFrame >= 65) {
+        if (currentTime - previousFrame >= oneSecond) {
```
Comment (platisd) [inline arduino/smartcar/smartcar.ino:88]: By doing this, your camera feed frame rate became 15 times slower. I don't see any good reason why it should be like that. Please revise.

File: arduino/smartcar/smartcar.ino [hunk 3]
```diff
@@ -85,20 +85,33 @@ void loop() {
...
+		  		  
+		  // Publish Camera data
           Camera.readFrame(frameBuffer.data());
-          mqtt.publish("/smartcar/camera", frameBuffer.data(), frameBuffer.size(),
-                       false, 0);
+          mqtt.publish("/smartcar/camera", frameBuffer.data(), frameBuffer.size(),false, 0);
+		  const auto totalDistance = String(travelledDistance());
+		  mqtt.publish("/smartcar/odometer", totalDistance);
+		  const auto distance = String(measureDistance());
+		  mqtt.publish("/smartcar/ultrasound/front", distance);
```
Comment (platisd) [inline arduino/smartcar/smartcar.ino:97]: Please indent the code appropriately. You can easily do this by pressing "Ctrl + T" on the Arduino IDE.

File: arduino/smartcar/smartcar.ino [hunk 4]
```diff
@@ -85,20 +85,33 @@ void loop() {
...
-        }
-      }
+          // mqtt.publish("/smartcar/ultrasound/front", distance);
+		  // Publish Total distance
+		  const auto totalDistance = String(travelledDistance());
+		  // Serial.println(totalDistance);		  
+		  //mqtt.publish("/smartcar/odometer", "11111");
+		  
+        }*/
+      }
```
Comment (platisd) [inline arduino/smartcar/smartcar.ino:114]: Commented out code is a code-smell. If you don't need it, remove it.

File: arduino/smartcar/smartcar.ino [hunk 5]
```diff
@@ -156,3 +169,7 @@ void rotateOnSpot(int targetDegrees) {
 int measureDistance() {
       return front.getDistance();
 }
+
+int travelledDistance() {
+      return leftOdometer.getDistance();
```
Comment (platisd) [inline arduino/smartcar/smartcar.ino:174]: Why only from the left odometer and not from both? The `DistanceCar` class (and therefore the `SmartCar` one too) has a function just for that.

**Review by platisd — CHANGES_REQUESTED** [review]
Please address the comments I left and add screenshots of the new functionality in the pull request body.

File: android/app/src/main/AndroidManifest.xml [hunk 6]
```diff
@@ -1,7 +1,8 @@
 <?xml version="1.0" encoding="utf-8"?>
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
     xmlns:tools="http://schemas.android.com/tools"
-    package="com.delirover.app">
+    package="com.delirover.app"
+    tools:ignore="MissingLeanbackLauncher">
```
Comment (platisd) [inline android/app/src/main/AndroidManifest.xml:5]: What is `MissingLeanbackLauncher` and why should it be ignored?

**Comment by platisd** [comment]
> guscuizi requested a review from platisd

I've already left my review. Please address the comments.

**Comment by guscuizi** [comment]
Hi,

I added this code by mistake, just delete it, it won’t affect the software function.
I will also delete it, in the project.

Sincerely,
Zizhuang Cui
获取 Outlook for iOS<https://aka.ms/o0ukef>
________________________________
发件人: Dimitris Platis ***@***.***>
发送时间: Saturday, July 31, 2021 12:20:09 AM
收件人: DIT112-V21/group-17 ***@***.***>
抄送: guscuizi ***@***.***>; Author ***@***.***>
主题: Re: [DIT112-V21/group-17] Odometer (#174)


@platisd commented on this pull request.

________________________________

In android/app/src/main/AndroidManifest.xml<https://github.com/DIT112-V21/group-17/pull/174#discussion_r680053118>:

> @@ -1,7 +1,8 @@
 <?xml version="1.0" encoding="utf-8"?>
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
     xmlns:tools="http://schemas.android.com/tools"
-    package="com.delirover.app">
+    package="com.delirover.app"
+    tools:ignore="MissingLeanbackLauncher">


What is MissingLeanbackLauncher and why should it be ignored?

―

**Comment by platisd** [comment]
> I added this code by mistake, just delete it, it won’t affect the software function.
I will also delete it, in the project.

First of all, I don't understand what code you are referring to. Please go to the [pull request on GitHub](https://github.com/DIT112-V21/group-17/pull/174) and respond to the comments there.
Secondly, this is a code review for **you** to fix the problems, not me. This is a process you should have already been familiar with...

**Comment by platisd** [comment]
Create another pull request towards the `master` branch so it can be properly reviewed please...

File: android/app/src/main/java/smart/delirover/app/MailmanTraveledDistanceActivity.java [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+    private static final int STRAIGHT_ANGLE = 0;
+    private static final int STEERING_ANGLE = 50;
+    private static final int QOS = 1;
+    private static final int IMAGE_WIDTH = 320;
+    private static final int IMAGE_HEIGHT = 240;
+
+    Context context;
+
+    
+    // I'm unfamilar with reusing the MqttClient, so I have no choice but to keep the code as it is.
```
Comment (platisd) [inline android/app/src/main/java/smart/delirover/app/MailmanTraveledDistanceActivity.java:47]: You already copy pasted the MqttClient, so it's difficult to see how "you have no choice" to reuse it. Please reuse it and don't respond with code comments to pull request review comments.

**Comment by guscuizi** [comment]
Okay, I will do this, thank you for your patience.

Sincerely，
ZiZhuang Cui
获取 Outlook for iOS<https://aka.ms/o0ukef>
________________________________
发件人: Dimitris Platis ***@***.***>
发送时间: Monday, August 2, 2021 4:13:26 PM
收件人: DIT112-V21/group-17 ***@***.***>
抄送: guscuizi ***@***.***>; State change ***@***.***>
主题: Re: [DIT112-V21/group-17] Odometer (#174)


Create another pull request towards the master branch so it can be properly reviewed please...

―

**Pull request merged**